src/main/java/net/sf/mzmine/modules/peaklistmethods/identification/ms2search/Ms2SearchTask.java

```diff
@@ -0,0 +1,280 @@
+        PeakListRow rows2[] = peakList2.getRows();
+        
+        int rows1Length = rows1.length;
+        int rows2Length = rows2.length;
+        
+        totalRows = rows1Length;
+        
+        for (int i = 0; i < rows1Length; i++) {
+            for (int j = 0; j < rows2Length; j++) {
+                Feature featureA = rows1[i].getBestPeak();
+                Feature featureB = rows2[j].getBestPeak();
+                
+                int featureAID = featureA.hashCode();
+                int featureBID = featureB.hashCode();
+                
+                searchResult = simpleMS2similarity(featureA,
+                        featureB, intensityThreshold, mzTolerance.getPpmTolerance());
+                
+                //Report the final score to the peaklist identity
+                if (searchResult != null && searchResult.getScore() > scoreThreshold && searchResult.getNumIonsMatched() >= minimumIonsMatched)
+                    addFragmentClusterIdentity(rows1[i],featureA,featureB,searchResult);
+                
+                //featureA.appendFragmentSimilarityScore(featureBID,result);
+              
+                if (isCanceled())
+                    return;
+            }
+            
+        //Update progress bar
+        finishedRows++;
+        }
+
+        // Add task description to peakList
+        ((SimplePeakList) peakList1).addDescriptionOfAppliedTask(
+                new SimplePeakListAppliedMethod("Identification of complexes",
+                        parameters));
+
+        // Repaint the window to reflect the change in the peak list
+        Desktop desktop = MZmineCore.getDesktop();
+        if (!(desktop instanceof HeadLessDesktop))
+            desktop.getMainWindow().repaint();
+
+        setStatus(TaskStatus.FINISHED);
+
+        logger.info("Finished MS2 similarity search for " + peakList1 + "against"
+                + peakList2);
+
+    }
+
+    /**
+     * Simple cosine like calculation of similarity between ions of two fragmentation scans.
+     * 
+     */
+    private Ms2SearchResult simpleMS2similarity(Feature featureA, Feature featureB,
+            double intensityThreshold, double mzRangePPM) {
+
+        double runningScoreTotal = 0.0;
+
+        // Fetch 1st peak MS2 scan.
+        int ms2ScanNumberA = featureA.getMostIntenseFragmentScanNumber();
+        Scan scanMS2A = featureA.getDataFile().getScan(ms2ScanNumberA);
+        RawDataFile featureADataFile = featureA.getDataFile();
+
+        // Fetch 2nd peak MS2 scan.
+        int ms2ScanNumberB = featureB.getMostIntenseFragmentScanNumber();
+        Scan scanMS2B = featureB.getDataFile().getScan(ms2ScanNumberB);
+        RawDataFile peak2DataFile = featureB.getDataFile();
+        
+        if (scanMS2A == null || scanMS2B == null)
+        {
+            return null;
+        }
+
+        DataPoint[] ionsA = null;
+        DataPoint[] ionsB = null;
+
+        MassList massListA = scanMS2A.getMassLists()[0];
```

Comment:
scanMS2A.getMassLists()[0] will throw an exception if there is no mass list at all. Plus, taking the first mass list might not always be optimal - why not add the MassListParameter to the parameter set and let the user choose which mass list to use.